layer/migration.go

```diff
@@ -178,6 +173,18 @@ func (ls *layerStore) RegisterByGraphID(graphID string, parent ChainID, tarDataF
 		referenceCount: 1,
 		layerStore:     ls,
 		references:     map[Layer]struct{}{},
+		diffID:         diffID,
+		size:           size,
+		chainID:        createChainIDFromParent(parent, diffID),
+	}
+
+	ls.layerL.Lock()
```

Comment:
I would just keep this whole block where it was. The case where the image already exists does mean that we have unnecessarily written out the metadata files but the transaction will clean that up. I would argue it is better to be optimistic here by writing out the files then potentially throwing them away at the end to reduce the time the layer lock is held. Acquiring the lock twice to avoid this just does not seem necessary.